src/server-glue/webresource-handler.ts

```diff
@@ -0,0 +1,344 @@
+				};
+				const promise = handler.handleFile(file).then((result) => {
+					req.body[fieldname] = {
+						filename: info.filename,
+						contentType: info.mimeType,
+						contentDisposition: undefined,
+						size: result.size,
+						href: result.filename,
+					};
+					filesUploaded.push(result.filename);
+				});
+				completeUploads.push(promise);
+			} else {
+				filestream.resume();
+			}
+		});
+
+		bb.on('field', (name, val, _info) => {
+			req.body[name] = val;
+		});
+
+		bb.on('finish', async () => {
+			try {
+				await Promise.all(completeUploads);
+				done();
+				next();
+			} catch (err: any) {
+				await clearFiles();
+
+				if (err instanceof FileSizeExceededError) {
+					return sbvrUtils.handleHttpErrors(
+						req,
+						res,
+						new errors.BadRequestError(err.message),
+					);
+				}
+
+				console.error('Error uploading file', err);
+				next(err);
+			}
+		});
+
+		bb.on('error', async (err) => {
+			await clearFiles();
+			done();
+			next(err);
+		});
+		req.pipe(bb);
+	};
+};
+
+// TODO: this can be cached
+const getWebResourceFields = (request: uriParser.ODataRequest): string[] => {
+	const fields =
+		request.abstractSqlModel?.tables[request.resourceName]?.modifyFields ??
+		request.abstractSqlModel?.tables[request.resourceName]?.fields;
+	if (fields == null) {
+		return [];
+	}
+
+	return fields
+		.filter((f) => f.dataType === 'WebResource')
+		.map((f) => sqlNameToODataName(f.fieldName));
+};
+
+const getModifiedFields = (request: uriParser.ODataRequest): string[] => {
+	return Object.entries(request.values)
+		.filter(([_key, value]) => value !== undefined)
+		.map(([key, _value]) => key);
+};
+
+const deleteFiles = async (
+	keysToDelete: string[],
+	webResourceHandler: WebResourceHandler,
+) => {
+	const promises = keysToDelete.map((r) => webResourceHandler.removeFile(r));
+	await Promise.all(promises);
+};
+
+const getCreateWebResourceHook = (webResourceHandler: WebResourceHandler) => {
+	return {
+		'POSTRUN-ERROR': async ({ tx, request }) => {
+			tx?.on('rollback', async () => {
+				const fields = getWebResourceFields(request);
+
+				if (fields.length === 0) {
+					return;
+				}
+
+				const keysToDelete: string[] = fields
+					.filter((f) => isDefined(request.values[f]))
+					.map((f) => request.values[f].href);
+				await deleteFiles(keysToDelete, webResourceHandler);
+			});
+		},
+	} as sbvrUtils.Hooks;
+};
+
+const isDefined = <T>(x: T | undefined | null): x is T => x != null;
+
+const getWebResourcesHrefs = (webResources?: WebResourcesDbResponse | null) => {
+	const hrefs = Object.values(webResources ?? {})
+		.filter(isDefined)
+		.map((resourceKey) => resourceKey.href);
+	return hrefs;
+};
+
+const getRemoveWebResourceHook = (webResourceHandler: WebResourceHandler) => {
+	return {
+		PRERUN: async (args) => {
+			const { api, request } = args;
+			let fields = getWebResourceFields(request);
+
+			if (request.method === 'PATCH') {
+				if (request.odataQuery?.key == null) {
```

Comment:
Shouldn't this check come after we have ensured that at least one field matches a webresource field?

Reply:
@Page- My understanding is that no. This is a limitation for fileuploads as it must have a odata key (e.g. you can't upload the same file to multiple resources at the same time). So it doesn't really matter who comes first here